### PR TITLE
PP-10715 DAO method to delete old idempotency keys

### PR DESCRIPTION
**Pull request by james-peacock-ho**
- Idempotency keys should be deleted when they are over 24 hours old
- Add DAO method to delete expiring keys
- Add integration test
- Add environment variable for idempotency key expiry threshold
- Changes to Resource and Service to call the deletion method will be added in a separate PR

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/idempotency/dao/IdempotencyDao.java [hunk 1]
```diff
@@ -25,4 +27,13 @@ public Optional<IdempotencyEntity> findByGatewayAccountIdAndKey(Long gatewayAcco
                 .setParameter("key", key)
                 .getResultList().stream().findFirst();
     }
+
+    public int deleteIdempotencyKeysMoreThan24HoursOld() {
+        Instant nowMinus24Hours = Instant.now().minus(1, ChronoUnit.DAYS);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/idempotency/dao/IdempotencyDao.java]: Can this be an env variable so it is easily configurable? especially when testing locally and we can also set different thresholds in other environments if we decide to.

Comment (james-peacock-ho) [reply src/main/java/uk/gov/pay/connector/idempotency/dao/IdempotencyDao.java]: updated as discussed

File: src/main/resources/config/config.yaml [hunk 2]
```diff
@@ -259,6 +259,7 @@ chargesSweepConfig:
   awaitingCaptureExpiryThreshold: ${AWAITING_DELAY_CAPTURE_EXPIRY_WINDOW:-432000}
   tokenExpiryThresholdInSeconds: ${TOKEN_EXPIRY_WINDOW_SECONDS:-604800}
   skipExpiringChargesLastUpdatedInSeconds: ${SKIP_EXPIRING_CHARGES_LAST_UPDATED_IN_SECONDS:-300}
+  idempotencyKeyExpiryThresholdInSeconds: ${IDEMPOTENCY_KEY_EXPIRY_THRESHOLD_IN_SECONDS:-86400}
```
Comment (kbottla) [inline src/main/resources/config/config.yaml:262]: Would be good to update test configs too to keep this in sync and config for integration tests https://github.com/alphagov/pay-connector/tree/master/src/test/resources/config. Can be in a separate PR.

File: src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java [hunk 3]
```diff
@@ -59,6 +59,32 @@ public void shouldThrowException_whenGatewayAccountIdAndKeyExists() {
         assertThrows(RollbackException.class, () -> dao.persist(entity));
     }
 
+    @Test
+    public void shouldDeleteOnlyIdempotencyEntitiesMoreThan24HoursOld() {
+        Map<String, Object> requestBody = Map.of("foo", "bar");
+        String expiringKey = "expiring-idempotency-key";
+        String newKey = "new-idempotency-key";
+        databaseTestHelper.insertIdempotency(expiringKey, gatewayAccount.getId(), resourceExternalId, requestBody);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java]: We can possibly override `insertIdempotency ` to insert a record with a created date and can avoid querying DB + updating record

https://github.com/alphagov/pay-connector/pull/4443/files#diff-88a66c9ac67ff4f4a879c9171a7bc2c691f37e2292de73ed7e602d1b1fe59c47R71-R74

File: src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java [hunk 4]
```diff
@@ -59,6 +59,32 @@ public void shouldThrowException_whenGatewayAccountIdAndKeyExists() {
...
+        databaseTestHelper.insertIdempotency(newKey, gatewayAccount.getId(), resourceExternalId, requestBody);
+
+        //backdate creation date of key to more than 24 hours ago
+        IdempotencyEntity idempotencyEntity = dao.findByGatewayAccountIdAndKey(gatewayAccount.getId(), expiringKey).get();
+        Instant nowMinus25Hours = Instant.now().minus(25, ChronoUnit.HOURS);
+        databaseTestHelper.updateIdempotencyCreatedDate(expiringKey, nowMinus25Hours);
+        IdempotencyEntity expiringIdempotencyEntity = dao.findByGatewayAccountIdAndKey(gatewayAccount.getId(), expiringKey).get();
+
+        IdempotencyEntity newIdempotencyEntity = dao.findByGatewayAccountIdAndKey(gatewayAccount.getId(), newKey).get();
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java]: Any reason these are queried and not used?

File: src/main/java/uk/gov/pay/connector/idempotency/dao/IdempotencyDao.java [hunk 5]
```diff
@@ -6,6 +6,9 @@
 import uk.gov.pay.connector.idempotency.model.IdempotencyEntity;
 
 import javax.persistence.EntityManager;
+import java.time.Instant;
+import java.time.ZonedDateTime;
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/idempotency/dao/IdempotencyDao.java]: Unused imports?

File: src/main/resources/config/config.yaml (hunk 2 above)
Comment (james-peacock-ho) [reply src/main/resources/config/config.yaml:262]: agreed, thanks

File: src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java (hunk 3 above)
Comment (james-peacock-ho) [reply src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java]: thanks - this has been changed as suggested

File: src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java (hunk 4 above)
Comment (james-peacock-ho) [reply src/test/java/uk/gov/pay/connector/it/dao/IdempotencyDaoIT.java]: no reason - my mistake. Now removed.

File: src/main/java/uk/gov/pay/connector/idempotency/dao/IdempotencyDao.java (hunk 5 above)
Comment (james-peacock-ho) [reply src/main/java/uk/gov/pay/connector/idempotency/dao/IdempotencyDao.java]: removed, thanks

**Review by kbottla — APPROVED** [review]
